crates/adkg/src/rbc/r4/handlers.rs

```diff
@@ -23,50 +24,70 @@ pub(super) async fn rbc_receive_proposal<T>(
 {
     let (n, t) = (st.n, st.t);
     let m = &proposal.m;
-    info!("Node `{}` received proposal from node `{sender}`", st.id);
     debug!(
         "Node `{}` received proposal from node `{sender}` with message m = `{m:?}`",
         st.id
     );
 
-    // if sent by the expected broadcaster AND P(M) then
+    if &sender != expected_sender {
+        // Not the expected sender
+        warn!(
+            ?expected_sender,
+            "Node `{}` refused proposal: unexpected sender", st.id
+        );
+        return;
+    }
+
+    if st.status != RbcStatus::WaitingForProposal {
+        // A proposal was already accepted
+        warn!(
+            "Node `{}` refused proposal: a proposal was already accepted",
+            st.id
+        );
+        return;
+    }
+
+    if !predicate.predicate(sender, m.as_slice()).await {
+        // Predicate refused the proposal
+        warn!(
+            "Node `{}` refused proposal: refused due to predicate",
+            st.id
+        );
+        return;
+    }
+
+    // We reach here if the proposal was sent by the expected broadcaster AND waiting for a proposal
+    // AND P(M) -- the predicate returned true
     // Notice that we halt processing of all upcoming messages until the predicate resolves.
-    if sender.eq(expected_sender)
-        && predicate.predicate(sender, m.as_slice()).await
-        && st.status == RbcStatus::WaitingForProposal
-    {
-        info!("Node `{}` accepted proposal", st.id);
+    info!("Node `{}` accepted proposal", st.id);
 
-        // let h := hash(M)
-        let h = Sha3_256::digest(m).to_vec();
-        // let M' = [m_1, m_2, ..., m_n] = RSEnc(M, n, t + 1)
-        let mp = rs_encode_stripes(m, n, t + 1);
+    // let h := hash(M)
+    let h = Sha3_256::digest(m).to_vec();
```

Comment:
Is this really meant to be hardcoded to be that hash rather than be generic coming from higher up in the scheme defs?